design/general-progress-monitoring.md

```diff
@@ -0,0 +1,496 @@
+    	// then velero will wait for this operation to complete before the restore is marked Completed.
+	Execute(input *RestoreItemActionExecuteInput) (*RestoreItemActionExecuteOutput, error)
+
+    	
+    	// Progress  
+    	Progress(input *RestoreItemActionProgressInput) (OperationProgress, error)
+
+        // Cancel  
+    	Cancel(input *RestoreItemActionProgressInput) error
+    
+    // RestoreItemActionExecuteOutput contains the output variables for the ItemAction's Execution function.
+    type RestoreItemActionExecuteOutput struct {
+        // UpdatedItem is the item being restored mutated by ItemAction.
+        UpdatedItem runtime.Unstructured
+
+        // AdditionalItems is a list of additional related items that should
+        // be restored.
+        AdditionalItems []ResourceIdentifier
+
+        // SkipRestore tells velero to stop executing further actions
+        // on this item, and skip the restore step. When this field's
+        // value is true, AdditionalItems will be ignored.
+        SkipRestore bool
+
+        // OperationID is an identifier which indicates an ongoing asynchronous action which Velero will
+        // continue to monitor after restoring this item. If left blank, then there is no ongoing operation
+        OperationID string
+    }
+
+    // RestoreItemActionProgressInput contains the input parameters for the RestoreItemAction's Progress function.
+    type RestoreItemActionProgressInput struct {
+    	// Item is the item that was stored in the restore
+    	Item runtime.Unstructured
+    	// OperationID is the operation ID returned by RestoreItemAction Execute
+    	operationID string
+    	// Restore is the representation of the restore resource processed by Velero.
+    	Restore *velerov1api.Restore
+    }
+
+## Changes in Velero backup format
+
+No changes to the existing format are introduced by this change.  A
+`<backup-name>-itemoperations.json.gz` file will be added that contains the items and operation IDs
+(snapshotIDs) returned by VolumeSnapshotter and BackupItemAction plugins.  Also, the creation of the
+`velero-backup.json` object will not occur until the backup moves to one of the terminal phases
+(_Completed_, _PartiallyFailed_, or _Failed_).  Reconciliation should ignore backups that do not
+have a `velero-backup.json` object.
+
+The cluster that is creating the backup will have the Backup resource present and will be able to
+manage the backup before the backup completes.
+
+If the Backup resource is removed (e.g. Velero is uninstalled) before a backup completes and writes
+its `velero-backup.json` object, the other objects in the object store for the backup will be
+effectively orphaned.  This can currently happen but the current window is much smaller.
+
+### `<backup-name>-itemoperations.json.gz`
+The itemoperations file is similar to the existing `<backup-name>-itemsnapshots.json.gz`  Each snapshot taken via
```

Comment:
Sure. I think, generally, this would be uploaded to the object store after processing all backup/restore items, as until this point we won't have the complete list. It will only be used on subsequent reconciles, so it won't be referenced until the backup moves to the next phase  beyond `InProgress`